Stop using StringView::operator[] in hot code paths in WebCore

cdumez opened the pull request:
#### 9ad44599784a7ab2e46317f371e0c33b667906fe
<pre>
Stop using StringView::operator[] in hot code paths in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=263216">https://bugs.webkit.org/show_bug.cgi?id=263216</a>

Reviewed by Darin Adler.

Stop using StringView::operator[] in hot code paths in WebCore. StringView&apos;s
operator[] is not super efficient because it needs to check the is8Bit flag for
every character when iterating. I am also looking into adding a boundary check
in StringView::operator[], which would further increase branching.

* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::moveToNextNonWhitespacePosition):
(WebCore::Layout::InlineItemsBuilder::handleTextContent):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::canUseSimplifiedTextMeasuringForCharacters):
(WebCore::Layout::TextUtil::canUseSimplifiedTextMeasuring):
* Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
(WebCore::Layout::canUseSimplifiedTextMeasuringForCharacters):
(WebCore::Layout::canUseSimplifiedTextMeasuring):

Canonical link: <a href="https://commits.webkit.org/269396@main">https://commits.webkit.org/269396@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e94b4da8beb002756b67736389423112cdb0a8cb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22411 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23490 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24315 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20744 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22665 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22944 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21747 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22650 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19456 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25169 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20311 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26558 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20394 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20549 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24413 "Found 1 new API test failure: /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/service-worker-registrations (failure)") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/17 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17874 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5342 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -43,7 +43,9 @@ struct WhitespaceContent {
-static std::optional<WhitespaceContent> moveToNextNonWhitespacePosition(StringView textContent, size_t startPosition, bool preserveNewline, bool preserveTab, bool stopAtWordSeparatorBoundary)
+
+template<typename CharacterType>
+static std::optional<WhitespaceContent> moveToNextNonWhitespacePosition(CharacterType* textContent, unsigned length, size_t startPosition, bool preserveNewline, bool preserveTab, bool stopAtWordSeparatorBoundary)
```
In new code I like to use `std::span<const CharacterType>`.

cdumez replied:
I purposefully didn't use std::span<> because its operator[] does a bounds check and this is hot code. 

Note that the reason I am moving this code away from StringView::operator[] is because adding such a bounds check to StringView::operator[] is a performance regression on both Jetstream and Speedometer.

Odds are that by using std::span here I'd introduce the same perf regression I saw when adding the bounds check to  StringView::operator[].

darinadler replied:
Unlikely `StringView`, there’s no reason we can’t have a fast unsafe index into a `span`, which can be written `span.data()[index]` or any other idiom we choose. But I suppose it’s a style choice. I think passing pairs of arguments as if they are unrelated rather than grouped in a `span` is obviously worse.

cdumez replied:
Yes, I could have passed a span and used span.data()[index] inside. I'll follow-up after I'm done with some urgent stuff.

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/19238.

darinadler commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -552,6 +552,20 @@ float TextUtil::hangableStopOrCommaEndWidth(const InlineTextItem& inlineTextItem
 }
 
+template<typename CharacterType>
+static bool canUseSimplifiedTextMeasuringForCharacters(const CharacterType* characters, unsigned length, const FontCascade& fontCascade, const Font& primaryFont, bool whitespaceIsCollapsed)
```
In new code I like to use `std::span<const CharacterType>`.

darinadler commented on Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
```diff
@@ -86,6 +86,20 @@ static std::optional<LayoutSize> accumulatedOffsetForInFlowPositionedContinuatio
 }
 
+template<typename CharacterType>
+static bool canUseSimplifiedTextMeasuringForCharacters(const CharacterType* characters, unsigned length, const FontCascade& fontCascade, bool whitespaceIsCollapsed)
```
In new code I like to use `std::span<const CharacterType>`.

cdumez committed:
Stop using StringView::operator[] in hot code paths in WebCore
https://bugs.webkit.org/show_bug.cgi?id=263216

Reviewed by Darin Adler.

Stop using StringView::operator[] in hot code paths in WebCore. StringView's
operator[] is not super efficient because it needs to check the is8Bit flag for
every character when iterating. I am also looking into adding a boundary check
in StringView::operator[], which would further increase branching.

* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::moveToNextNonWhitespacePosition):
(WebCore::Layout::InlineItemsBuilder::handleTextContent):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::canUseSimplifiedTextMeasuringForCharacters):
(WebCore::Layout::TextUtil::canUseSimplifiedTextMeasuring):
* Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
(WebCore::Layout::canUseSimplifiedTextMeasuringForCharacters):
(WebCore::Layout::canUseSimplifiedTextMeasuring):

Canonical link: https://commits.webkit.org/269396@main
